### DISCUSSION TITLE
Generic Crypto Implementation

### DISCUSSION THREAD

**Discussion by Mirio** [ghostfolio/ghostfolio#1668] [Q&A]
Hi All,
Could you add crypto integration? Currently, allow only 3-4 "most used" coins.

I can suggest using the coin gecko API because:
* Doesn't require any registration to use the API
* Is not related to any specific crypto exchange
* Have a large rate limit
* A large set of coins supported

**Rate limits**: 10-50 calls/minute
**Docs**: [https://www.coingecko.com/en/api/documentation](https://www.coingecko.com/en/api/documentation)

## Coin List
**Request**
`curl -X 'GET'  'https://api.coingecko.com/api/v3/coins/list'  -H 'accept: application/json'`

**Response**:
`[
  {
    "id": "01coin",
    "symbol": "zoc",
    "name": "01coin"
  },
  {
    "id": "0chain",
    "symbol": "zcn",
    "name": "Zus"
  },
  {
    "id": "0x",
    "symbol": "zrx",
    "name": "0x"
  }
...]`

## Simple price:
**Request**:
`curl -X 'GET' 'https://api.coingecko.com/api/v3/simple/price?ids=bitcoin&vs_currencies=eur'  -H 'accept: application/json'`

**Response**:
`{
  "bitcoin": {
    "eur": 21224
  }
}`

**Comment by dtslvr** [comment]
Hi @Mirio

Is your question related to self-hosting or the cloud offering?

Ghostfolio supports uncountable cryptocurrencies. Which cryptocurrency do you miss in particular?

**Reply by Mirio** [reply]
Hi @dtslvr,
I use the latest self hosted version (v1.230.0).
**Just an example from a random walk**: CRO (Cronos), USDT (Tether), TON (Toncoin), HOT (holo), COMP (Compound), etc...

**Additional note**: Currently Ghostfolio supported only cryptocurrency VS USD using those api there would be no conversion problem.

Thanks

**Reply by dtslvr** [reply]
I was able to track _Cronos_ and _Tether_ in the cloud.

The limitation to `USD` is a valid point though. Are you interested in implementing a data provider for _CoinGecko_? You can find an example here: [EodHistoricalDataService](https://github.com/ghostfolio/ghostfolio/blob/main/apps/api/src/services/data-provider/eod-historical-data/eod-historical-data.service.ts).

The most important functions are `getHistorical()`, `getQuotes()` and `search()`.

**Reply by Mirio** [reply]
Sure i can try it!
Did you have some "contribution start guide" or something like that?

**Reply by dtslvr** [reply]
You can find a guide to setup the project [here](https://github.com/ghostfolio/ghostfolio#development).

If you have any questions about setting it up or with the implementation, it's best to get in touch on [Slack](https://ghostfolio.slack.com).
